### PR TITLE
build: upgrade ibc-go from v2 to v3

### PR DESCRIPTION
**Pull request by queencre**
## Description

~This PR migrates from ibc-go from [v2.2.0](https://github.com/cosmos/ibc-go/tree/v2.2.0) to either [v3.3.1](https://github.com/cosmos/ibc-go/tree/v3.3.1).~

This PR migrates from ibc-go from [v2.2.0](https://github.com/cosmos/ibc-go/tree/v2.2.0) to either [v3.4.0](https://github.com/cosmos/ibc-go/tree/v3.4.0).

closes: #124 

## Tasks

- [x] Bump ibc-go and cosmos-sdk
- [x] Add v4 upgrade handler script (Allow `MsgLimitOrder` and `MsgMarketOrder` custom messages)
- [x] Commend out some lines of code in the `farming` module due to reverted customization in cosmos-sdk (dynamic block address)
- [x] Resolve golangci linter issues
- [x] Resolve failing test cases
- [x] Update docs

## References

- [Migrating from ibc-go from v2 to v3](https://github.com/cosmos/ibc-go/blob/v3.3.1/docs/migrations/v2-to-v3.md)
- [Migrating from not supporing base denoms with slashes to supporting base denoms with slashes](https://github.com/cosmos/ibc-go/blob/v2.4.0/docs/migrations/support-denoms-with-slashes.md)

## Notes

Transferring tokens with slashes from Chain A (ibc-go v3.4.0) to Chain B (ibc-go v2.2.0) fails due to the following error message `ABCI code: 1: error handling packet on destination chain: see events for details`. See [this transaction hash](https://www.mintscan.io/cosmos/txs/B3FE9F5605BC79F5C50EABB6E94DF943BD89F89B97F8385DC60444CC57786EA9) that is executed to try transferring tokens with slashes from Chain A (ibc-go v3.4.0) to Cosmos Hub (ibc-go v3.0.0 at the time of writing).

After doing some research, ABCI code 1 is an internal error and it is coming from `trace info must come in pairs of port and channel identifiers` error in [this line of code](https://github.com/cosmos/ibc-go/blob/46e020640e66f9043c14c53a4d215a5b457d6703/modules/apps/transfer/types/trace.go#L75) because the IBC `transfer` module expects trace path information to always have an even number of identifiers. It's unfortunate that the events don't tell much about the actual error. Error message handling is improved in v3.4.0 and that is why we need to upgrade Core asap.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Appropriate labels applied
- [ ] Targeted PR against correct branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/v0.45.9/docs/building-modules/README.md).
- [ ] Wrote unit and integration
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://go.dev/blog/godoc).
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

**Comment by queencre** [comment]
We can discuss if we need to upgrade `ibc-go` to v3.4.0 or even above (v0.4.x or v0.5.x),.

File: app/app.go [hunk 1]
```diff
@@ -531,22 +547,38 @@ func NewApp(
 		app.LiquidityKeeper,
 		app.LiquidStakingKeeper,
 	)
-
 	app.TransferKeeper = ibctransferkeeper.NewKeeper(
 		appCodec,
 		keys[ibctransfertypes.StoreKey],
 		app.GetSubspace(ibctransfertypes.ModuleName),
 		app.IBCKeeper.ChannelKeeper,
+		app.IBCKeeper.ChannelKeeper,
```
Comment (richard-bachman) [inline app/app.go:555]: This line seems the same as the right above one. Could you check this?

**Review by richard-bachman — COMMENTED** [review]
Thanks a lot for your work. In task, there is `Add v4 upgrade handler script (Allow MsgLimitOrder and MsgMarketOrder custom messages)`. Shouldn't this be out?

File: app/app_test.go [hunk 2]
```diff
@@ -174,32 +174,33 @@ func TestRunMigrations(t *testing.T) {
...
-					"farming":       farming.AppModule{}.ConsensusVersion(),
-					"liquidity":     liquidity.AppModule{}.ConsensusVersion(),
-					"liquidstaking": liquidstaking.AppModule{}.ConsensusVersion(),
-					"liquidfarming": liquidfarming.AppModule{}.ConsensusVersion(),
-					"claim":         claim.AppModule{}.ConsensusVersion(),
-					"marketmaker":   marketmaker.AppModule{}.ConsensusVersion(),
-					"lpfarm":        lpfarm.AppModule{}.ConsensusVersion(),
-					"ibc":           ibc.AppModule{}.ConsensusVersion(),
-					"transfer":      transfer.AppModule{}.ConsensusVersion(),
+					"bank":               1,
```
[the remark this answers is not in the document]
Comment (queencre) [reply app/app_test.go:177]: If you look at the test case (little above), it explains why bank consensus version should be 1.

**Comment by queencre** [comment]
>  In task, there is Add v4 upgrade handler script (Allow MsgLimitOrder and MsgMarketOrder custom messages). Shouldn't this be out?

I'm not sure if I follow. If you look inside the `upgrades/mainnet/v4/upgrades.go`, we should define what messages to be allowed as a host chain.

File: go.mod [hunk 3]
```diff
@@ -3,23 +3,23 @@ module github.com/crescent-network/crescent/v3
 go 1.18
 
 require (
-	github.com/cosmos/cosmos-sdk v0.45.9
-	github.com/cosmos/ibc-go/v2 v2.2.0
+	github.com/cosmos/cosmos-sdk v0.45.10
```
Comment (kingcre) [inline go.mod:6]: Do we need to bump Cosmos SDK to `v0.45.10` as well in this patch? The `replace` directive in this file specifies `github.com/cosmos/cosmos-sdk => github.com/crescent-network/cosmos-sdk v1.1.3-sdk-0.45.9`

Comment (queencre) [reply go.mod:6]: Will upgrade Cosmos SDK from v0.45.9 to v0.45.10 now and apply it to this PR.

Comment (queencre) [reply go.mod:7]: Applied.

**Review by kingcre — APPROVED** [review]
LGTM

**Pull request merged**